File: hackerrank/typescript/src/the-hurdle-race.ts
/**
 * @param k - Initial maximum height one can jump
 * @param height - Array of heights of hurdles
 * @returns Number of dose
 *
 * Challenge
 * https://www.hackerrank.com/challenges/the-hurdle-race/problem
 */
export const hurdleRace = (k: number, height: number[]): number => {
	const max = Math.max.apply(null, height)

	return k >= max ? 0 : max - k
}

if (import.meta.vitest) {
	const { it, expect } = import.meta.vitest

	it('Hurdle Race', () => {
		expect(hurdleRace(1, [1, 2, 3, 3, 4])).toBe(2)
	})
}

File: hackerrank/rust/src/the_hurdle_race.rs
#[allow(dead_code)]
/// Check how many doses one needs to jump all hurdles
///
/// * `k`: The maxium height one can jump
/// * `height`: Array of heights of hurdles
///
/// Challenge:
/// https://www.hackerrank.com/challenges/the-hurdle-race/problem

fn hurdle_race(k: i32, height: &[i32]) -> i32 {
    let max = height.iter().max();
    match max {
        Some(&h) if h > k => h - k,
        Some(&h) if h <= k => 0,
        _ => panic!("Error"),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_one() {
        assert_eq!(hurdle_race(1, &[1, 2, 3, 3, 2]), 2);
    }
}
